### PR TITLE
Modsourman 9 Initializing job executions

### PR DESCRIPTION
**Pull request by Igor-Gorchakov**
- REST method receives InitJobExecutionsRqDto with files,
- Service method does parentJobExecutionId generation,
- Service method prepares a necessary amount of JobExecution entities,
- Service method prepares a necessary amount Snapshot json entities,
- Service method does save prepared JobExecution entities using DAO,
- Service method calls mod-source-record-storage to save prepared Snapshots,
- Service method completes future and returns InitJobExecutionsRsDto with parentJobExecutionId and prepared JobExecution entities,
- REST method returns response containing InitJobExecutionsRsDto

### CODE REVIEW

File: src/main/java/org/folio/services/JobExecutionService.java [hunk 1]
```diff
@@ -21,4 +25,12 @@
    */
   Future<JobExecutionCollectionDto> getCollectionDtoByQuery(String query, int offset, int limit);
 
+  /**
+   * Performs creation of JobExecution and Snapshot entities
+   *
+   * @param dto Dto contains request params enough to create JobExecution and Snapshot entities
+   * @return Future
+   */
+  Future<InitJobExecutionsRsDto> initializeJobExecutions(InitJobExecutionsRqDto dto, OkapiConnectionParams params);
```
Comment (Igor-Gorchakov) [inline src/main/java/org/folio/services/JobExecutionService.java]: Maybe it's better to rename this method at all, some like initializeImport or Somethin' Else

File: src/main/java/org/folio/services/JobExecutionServiceImpl.java [hunk 2]
```diff
@@ -31,4 +51,109 @@ public JobExecutionServiceImpl(Vertx vertx, String tenantId) {
...
+    } else {
+      String parentJobExecutionId = UUID.randomUUID().toString();
+
+      List<JobExecution> jobExecutions = prepareJobExecutionList(parentJobExecutionId, jobExecutionsRqDto.getFiles());
+      List<JsonObject> snapshots = prepareSnapshotList(jobExecutions);
+
+      saveJobExecutions(jobExecutions);
+      saveSnapshots(snapshots, params);
+
+      // TODO wait for the completion of save JobExecutions and Snapshots if needed
```
Comment (Igor-Gorchakov) [inline src/main/java/org/folio/services/JobExecutionServiceImpl.java]: Now I'm adding ComposableFuture to handle this completion

File: src/main/java/org/folio/util/RestUtil.java [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+import io.vertx.core.http.HttpClientResponse;
+import io.vertx.core.http.HttpMethod;
+import io.vertx.core.json.JsonObject;
+
+import java.util.Map;
+
+/**
+ * Util class with static method for sending http request
+ */
+public class RestUtil {
```
Comment (Igor-Gorchakov) [inline src/main/java/org/folio/util/RestUtil.java]: Copied from mod-data-import

File: src/main/java/org/folio/util/OkapiConnectionParams.java [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+import java.util.Map;
+
+import static org.folio.util.RestUtil.OKAPI_TENANT_HEADER;
+import static org.folio.util.RestUtil.OKAPI_TOKEN_HEADER;
+import static org.folio.util.RestUtil.OKAPI_URL_HEADER;
+
+/**
+ * Wrapper class for Okapi connection params
+ */
+public class OkapiConnectionParams {
```
Comment (Igor-Gorchakov) [inline src/main/java/org/folio/util/OkapiConnectionParams.java]: Stolen from mod-data-import

File: src/main/java/org/folio/util/RestUtil.java [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+
+  public static final String OKAPI_TENANT_HEADER = "x-okapi-tenant";
+  public static final String OKAPI_TOKEN_HEADER = "x-okapi-token";
+  public static final String OKAPI_URL_HEADER = "x-okapi-url";
+  public static final int CREATED_STATUS_CODE = 201;
+
+  public static class WrappedResponse {
```
Comment (Igor-Gorchakov) [inline src/main/java/org/folio/util/RestUtil.java]: Stolen from mod-data-import

File: src/main/java/org/folio/services/JobExecutionServiceImpl.java [hunk 6]
```diff
@@ -31,4 +51,107 @@ public JobExecutionServiceImpl(Vertx vertx, String tenantId) {
...
+  private List<JobExecution> prepareJobExecutionList(String parentJobExecutionId, List<File> files) {
+    List<JobExecution> result = new ArrayList<>();
+    if (files.size() > 1) {
+      for (File file : files) {
+        JobExecution child = new JobExecution()
+          .withId(UUID.randomUUID().toString())
+          .withParentJobId(parentJobExecutionId)
+          .withSubordinationType(JobExecution.SubordinationType.CHILD)
+          .withStatus(JobExecution.Status.NEW)
+          .withSourcePath(file.getName());
```
Comment (Igor-Gorchakov) [inline src/main/java/org/folio/services/JobExecutionServiceImpl.java]: This is a temporary solution to set file name to sourcePath

File: src/main/java/org/folio/services/JobExecutionServiceImpl.java [hunk 7]
```diff
@@ -31,4 +51,145 @@ public JobExecutionServiceImpl(Vertx vertx, String tenantId) {
         .withTotalRecords(jobExecutionCollection.getTotalRecords()));
   }
 
+  public Future<InitJobExecutionsRsDto> initializeJobExecutions(InitJobExecutionsRqDto jobExecutionsRqDto, OkapiConnectionParams params) {
+    if (jobExecutionsRqDto.getFiles().isEmpty()) {
+      logger.error("Received files must not be empty");
+      return Future.failedFuture(new BadRequestException());
```
Comment (KaterynaSenchenko) [inline src/main/java/org/folio/services/JobExecutionServiceImpl.java]: Please add a message to the exception

File: src/main/java/org/folio/services/JobExecutionService.java [hunk 8]
```diff
@@ -21,4 +25,12 @@
    */
   Future<JobExecutionCollectionDto> getCollectionDtoByQuery(String query, int offset, int limit);
 
+  /**
+   * Performs creation of JobExecution and Snapshot entities
+   *
+   * @param dto Dto contains request params enough to create JobExecution and Snapshot entities
```
Comment (KaterynaSenchenko) [inline src/main/java/org/folio/services/JobExecutionService.java]: Please correct the javadoc to include all params

File: src/main/java/org/folio/rest/impl/ChangeManagerImpl.java [hunk 9]
```diff
@@ -4,36 +4,37 @@
...
-          );
-        }
-        asyncResultHandler.handle(Future.succeededFuture(PostChangeManagerJobExecutionsResponse.respond201WithApplicationJson(rsDto)));
+        OkapiConnectionParams params = new OkapiConnectionParams(okapiHeaders, vertxContext.owner());
+        jobExecutionService.initializeJobExecutions(initJobExecutionsRqDto, params)
+          .map(initJobExecutionsRsDto ->
+            (Response) PostChangeManagerJobExecutionsResponse.respond201WithApplicationJson(initJobExecutionsRsDto))
+          .otherwise(ExceptionHelper::mapExceptionToResponse)
+          .setHandler(asyncResultHandler);
       } catch (Exception e) {
```
Comment (KaterynaSenchenko) [inline src/main/java/org/folio/rest/impl/ChangeManagerImpl.java]: Please add logging

**Pull request merged**